lib/Dialect/FIRRTL/Transforms/ExtractInstances.cpp

```diff
@@ -0,0 +1,409 @@
+//===- ExtractInstances.cpp - Move instances up the hierarchy ---*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// Moves annotated instances upwards in the module hierarchy. Corresponds to the
+// `ExtractBlackBoxes`, `ExtractClockGates`, and `ExtractSeqMems` passes in the
+// Scala FIRRTL implementation.
+//
+//===----------------------------------------------------------------------===//
+
+#include "PassDetails.h"
+#include "circt/Dialect/FIRRTL/AnnotationDetails.h"
+#include "circt/Dialect/FIRRTL/FIRRTLAnnotations.h"
+#include "circt/Dialect/FIRRTL/FIRRTLInstanceGraph.h"
+#include "circt/Dialect/FIRRTL/Passes.h"
+#include "circt/Dialect/HW/HWAttributes.h"
+#include "circt/Dialect/HW/HWDialect.h"
+#include "circt/Dialect/SV/SVOps.h"
+#include "circt/Support/Path.h"
+#include "mlir/IR/Attributes.h"
+#include "mlir/IR/ImplicitLocOpBuilder.h"
+#include "mlir/Support/FileUtilities.h"
+#include "llvm/ADT/SmallPtrSet.h"
+#include "llvm/Support/Debug.h"
+#include "llvm/Support/MemoryBuffer.h"
+#include "llvm/Support/Path.h"
+
+#define DEBUG_TYPE "firrtl-extract-instances"
+
+using namespace circt;
+using namespace firrtl;
+
+//===----------------------------------------------------------------------===//
+// Pass Implementation
+//===----------------------------------------------------------------------===//
+
+namespace {
+/// All information necessary to move instances about.
+struct ExtractionInfo {
+  /// A filename into which the performed hierarchy modifications are emitted.
+  StringRef traceFilename;
+  /// A prefix to attach to the wiring generated by the extraction.
+  StringRef prefix;
+  /// Optional name of the wrapper module that will hold the moved instance.
+  StringRef wrapperModule;
+  /// The target module for the upwards move. This is strictly a parent of the
+  /// operation being moved.
+  Operation *upwardsTarget = nullptr;
+};
+
+struct ExtractInstancesPass
+    : public ExtractInstancesBase<ExtractInstancesPass> {
+  void runOnOperation() override;
+  void collectAnnos();
+  void collectAnno(InstanceOp inst, Annotation anno);
+  void extractInstances();
+
+private:
+  bool anythingChanged;
+  bool anyFailures;
+
+  InstanceGraph *instanceGraph = nullptr;
+
+  /// The target directory to output instances into. Can be changed through
+  /// `firrtl.transforms.BlackBoxTargetDirAnno` annotations.
+  StringRef targetDir;
+
+  /// The modules in the design that are annotated with one or more annotations
+  /// relevant for instance extraction.
+  DenseMap<Operation *, SmallVector<Annotation, 1>> annotatedModules;
+
+  /// All modules that are marked as DUT themselves, or that have a DUT parent
+  /// module.
+  DenseSet<Operation *> dutModules;
```

Comment:
Is the reason why this is being used, as opposed to just `InstanceGraph::isAncestor` to avoid the expense of `isAncestor` which does no caching?

Reply:
Yeah totally. The hash set does almost-constant lookup while the InstanceGraph does a linear scan up the hierarchy. Probably totally esoteric, and the linear scan might be faster given the fact that hierarchies are probably a dozen modules deep, and the hashing code probably takes a similar amount of time than traversing those modules :see_no_evil:.

This felt easier since it doesn't require me emitting errors if multiple things are marked as DUT and the like, and gives that faster lookup at first glance.